Restructure the abstract and overview (#4329)
quicwg/base-drafts: draft-ietf-quic-transport.md

--- abstract

This document defines the core of the QUIC transport protocol. QUIC provides
multiplexed streams for delivering ordered application data, flow control,

> low-latency connection establishment, migration, resilience to path changes,

Reviewer: For the abstract, I feel like it's sufficient to let migration and path changes be the same thing.

Author: Agreed.  I've tweaked this.  And the next sentence too, mainly to avoid a possessive.

--- abstract

This document defines the core of the QUIC transport protocol. QUIC provides

> streams for structuring application data, flow control,

Reviewer 2: structuring application data is not clear to me.

```suggestion
flow-controlled streams for transmitting application data,
```

Reviewer 3: I don't think you want to remove flow control from the connection-level descriptors of QUIC.

On "structuring data", the idea is that streams are a way to retain and structure application data. How about "QUIC provides streams for retaining structure in application data, flow control" ?

Reviewer 2: Streams are flow controlled, but other frames are not, so I was attempting to connect the two.  Saying you have flow control and not indicating what you're controlling seems potentially confusing to me.

Author: I think that I prefer Ian's suggestion here.  Breaking these apart doesn't make sense any more.

This could be two sentences instead: one for streams, one for the prime characteristics.  But then Ian's suggestion is shorter than my original proposal.

Author: > QUIC provides applications with flow-controlled streams for structured communication, low-latency connection establishment, and network path migration.  QUIC includes security measures that ensure confidentiality, integrity, and availability in a range of deployment circumstances.

^^ ?

Reviewer 3: I like your proposed text @martinthomson

> multiplexed streams for delivering ordered application data, flow control,

Reviewer 3: maybe?
```suggestion
streams for structuring application data, flow control,
```

Reviewer 3: I don't think "delivering ordered data" is quite right... at best this should be "ordered delivery of application data". But that is a tricky one, since there's no ordering across streams. Hence my suggestion of streams as something that an application uses for structuring data.

Also, I think "multiplexed" is an unnecessary detail for an overview.

This document defines the core of the QUIC transport protocol. QUIC provides
multiplexed streams for delivering ordered application data, flow control,
low-latency connection establishment, network path migration, and uses
authenticated encryption for most protocol data. Accompanying documents describe

> the integration of TLS for key negotiation, loss detection and an exemplary

Reviewer: ```suggestion
the integration of TLS for key negotiation, loss detection, and an exemplary
```
Oxford comma purist, checking in.  😁

# Overview

> QUIC is a multiplexed and secure general-purpose transport protocol. This

Reviewer 3: ```suggestion
QUIC is a secure general-purpose transport protocol. This
```

Reviewer 3: "multiplexed" is an odd qualifier here.

document defines version 1 of QUIC, which conforms to the version-independent
properties of QUIC defined in {{QUIC-INVARIANTS}}.

QUIC is a connection-oriented protocol that creates a stateful interaction
between a client and server.

The QUIC handshake combines negotiation of cryptographic and transport

> parameters. QUIC integrates the TLS {{?TLS13}} handshake, though provides a

Reviewer 3: ```suggestion
parameters. QUIC integrates the TLS ({{?TLS13}}) handshake, although using a
```

> customized scheme for protecting packets. The integration of TLS and QUIC is

Reviewer 3: Something better than "scheme"? Maybe "framing"?
```suggestion
customized framing for protecting packets. The integration of TLS and QUIC is
```

described in more detail in {{QUIC-TLS}}. The handshake is structured to permit
the exchange of application data as soon as possible. This includes an option
for clients to send data immediately (0-RTT), which might require prior
communication to enable.

Endpoints communicate in QUIC by exchanging QUIC packets. Most packets contain

> frames, which carry both control information and application data between

Reviewer 2: To me, "both" makes implies that one frame carries both control and application data.

```suggestion
frames, which carry control information and application data between
```

> frames, which carry both control information and application data between

Reviewer 3: ```suggestion
frames, which carry both control information and application data between the
```

credit-based scheme is used to limit stream creation and to bound the amount of
data that can be sent.

Application data can be reliably delivered by QUIC. An acknowledgment-based
algorithm for detecting and recovering from loss of data is described in
{{QUIC-RECOVERY}}. QUIC depends on congestion control to avoid network
congestion. An exemplary congestion control algorithm is also described in

> {{QUIC-RECOVERY}}.

Reviewer: I feel like it's worth mentioning that QUIC provides the feedback necessary to implement such algorithms, but doesn't mandate a particular one.

Author: I tried, but couldn't find anything satisfactory.  I don't think that ECN needs top-billing any more than the spin bit or any of the other minor protocol features we have.

Reviewer: I'm not suggesting calling out ECN specifically, but making sure the general focus of -transport versus -recovery is clear.  Transport defines feedback mechanisms, but no algorithm.  Recovery provides *an* algorithm, but not *the* algorithm.

Perhaps as simple as:

> QUIC provides the necessary feedback to implement reliable delivery and congestion control. An algorithm for detecting and recovering from loss of data is described in....

Author: OK, now I think that I get it.  And I like it.  Thanks.
